src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs

```diff
@@ -0,0 +1,2984 @@
+        this = new S<T> { F1 = t0 };
+        this = default;
+        S<T> s;
+        s = new S<T>();
+        s = new S<T>(ref t0);
+        s = new S<T> { F1 = t0 };
+    }
+    static void M1(T t1)
+    {
+        S<T> s1;
+        s1 = default;
+        s1 = new S<T>();
+        s1 = new S<T>(ref t1);
+        s1 = new S<T> { F1 = t1 };
+    }
+    static void M2(ref T t2)
+    {
+        S<T> s2;
+        s2 = new S<T>(ref t2);
+        s2 = new S<T> { F1 = t2 };
+    }
+    static void M3(S<T> s3)
+    {
+        var other = s3;
+        M1(s3.F1);
+        M1(s3.F2);
+        M2(ref s3.F1);
+    }
+    void M4(T t4)
+    {
+        this = default;
+        this = new S<T>();
+        this = new S<T>(ref t4);
+        this = new S<T> { F1 = t4 };
+        S<T> s;
+        s = new S<T>();
+        s = new S<T>(ref t4);
+        s = new S<T> { F1 = t4 };
+    }
+    void M5(S<T> s5)
+    {
+        var other = this;
+        M1(F1);
+        M1(F2);
+        M2(ref F1);
+        M1(this.F1);
+        M1(this.F2);
+        M2(ref this.F1);
+        M1(s5.F1);
+        M1(s5.F2);
+        M2(ref s5.F1);
+    }
+}";
+            var comp = CreateCompilation(sourceA, parseOptions: TestOptions.Regular10);
+            comp.VerifyEmitDiagnostics(
+                // (3,12): error CS8652: The feature 'ref fields' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //     public ref T F1;
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "ref T").WithArguments("ref fields").WithLocation(3, 12),
+                // (4,12): error CS8652: The feature 'ref fields' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //     public ref readonly T F2;
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "ref readonly T").WithArguments("ref fields").WithLocation(4, 12));
+
+            comp = CreateCompilation(sourceA);
+            comp.VerifyEmitDiagnostics();
+            var refA = AsReference(comp, useCompilationReference);
+
+            var sourceB =
+@"class Program
+{
+    static void M1<T>(T t)
+    {
+        S<T> s1;
+        s1 = default;
+        s1 = new S<T>();
+        s1 = new S<T>(ref t);
+        s1 = new S<T> { F1 = t };
+    }
+    static void M2<T>(ref T t)
+    {
+        S<T> s2;
+        s2 = new S<T>(ref t);
+        s2 = new S<T> { F1 = t };
+    }
+    static void M3<T>(S<T> s)
+    {
+        var s3 = s;
+        M1(s.F1);
+        M1(s.F2);
+        M2(ref s.F1);
+    }
+}";
+
+            comp = CreateCompilation(sourceB, references: new[] { refA }, parseOptions: TestOptions.Regular10);
+            comp.VerifyEmitDiagnostics(
+                // (9,25): error CS8652: The feature 'ref fields' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         s1 = new S<T> { F1 = t };
```

Comment:
Why is this a diagnostic but the same code executed inside `refA` is not?

Reply:
We reporting the `-langversion` error for use of `ref` fields when the field is declared outside the source module only. For fields declared in the same source module, we report a `-langversion` error on the field _declaration_ rather than the use. (Reporting on the use is redundant, and also a potential source of binding cycles.)